src/server.rs

```diff
@@ -375,7 +392,7 @@ impl<T, B: IntoBuf> Future for Handshake<T, B> {
 }
 
 impl<T, B> fmt::Debug for Handshake<T, B>
-    where T: fmt::Debug,
+    where T: AsyncRead + AsyncWrite + fmt::Debug,
```

Comment:
I think this bound could be removed as well.

Reply:
The bound is required by `Handshake<T: AsyncRead + AsyncWrite, B: IntoBuf>` now, since the `AsyncRead + AsyncWrite` bound is required by the implementation of `Future` for `ReadPreface` (which is then required by the `ReadPreface` type parameter on `future::AndThen` in `Handshake`.

Reply:
Double-checked this to make sure, and yeah, the compiler will not permit the bound to be removed:
<details>
<summary>
cargo build output
</summary>

```rust
   Compiling http v0.1.0 (https://github.com/carllerche/http?rev=2dd15d9#2dd15d9d)
   Compiling h2 v0.1.0 (file:///Users/eliza/Code/h2)
error[E0277]: the trait bound `T: tokio_io::AsyncRead` is not satisfied
   --> src/server.rs:394:12
    |
394 | impl<T, B> fmt::Debug for Handshake<T, B>
    |            ^^^^^^^^^^ the trait `tokio_io::AsyncRead` is not implemented for `T`
    |
    = help: consider adding a `where T: tokio_io::AsyncRead` bound
    = note: required by `server::Handshake`

error[E0277]: the trait bound `T: tokio_io::AsyncWrite` is not satisfied
   --> src/server.rs:394:12
    |
394 | impl<T, B> fmt::Debug for Handshake<T, B>
    |            ^^^^^^^^^^ the trait `tokio_io::AsyncWrite` is not implemented for `T`
    |
    = help: consider adding a `where T: tokio_io::AsyncWrite` bound
    = note: required by `server::Handshake`

error[E0277]: the trait bound `T: tokio_io::AsyncRead` is not satisfied
   --> src/server.rs:398:5
    |
398 | /     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
399 | |         write!(fmt, "server::Handshake")
400 | |     }
    | |_____^ the trait `tokio_io::AsyncRead` is not implemented for `T`
    |
    = help: consider adding a `where T: tokio_io::AsyncRead` bound
    = note: required by `server::Handshake`

error[E0277]: the trait bound `T: tokio_io::AsyncWrite` is not satisfied
   --> src/server.rs:398:5
    |
398 | /     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
399 | |         write!(fmt, "server::Handshake")
400 | |     }
    | |_____^ the trait `tokio_io::AsyncWrite` is not implemented for `T`
    |
    = help: consider adding a `where T: tokio_io::AsyncWrite` bound
    = note: required by `server::Handshake`

error: aborting due to 4 previous errors

error: Could not compile `h2`.
```
</details>

Reply:
Ok... I think it would be worth trying to figure out if there is a route by which the bound could be avoided... I don't know off the top of my head how.

Reply:
I'm pretty sure if we want to have the unboxed future in the `Handshake` struct, there's no way to get rid of the `AsyncRead + AsyncWrite` bound on the struct (without removing it from `impl Future for ReadPreface`, which seems unlikely).

Reply:
FWIW, the `client::Handshake` already has these bounds...

Reply:
Ok, I thought it didn't? Maybe it was added more recently.

Anyway, I'm fine w/ this. It would be good to merge. We can audit bounds when we do a pre 0.1 API audit pass.